server/app/services/program/ProgramServiceImpl.java

```diff
@@ -218,14 +218,67 @@ public ErrorAnd<ProgramDefinition, CiviFormError> updateProgramDefinition(
             .join());
   }
 
+  private boolean localizationStatusesAreInSyncWithConfiguredStatuses(
+      LocalizationUpdate localizationUpdate, ProgramDefinition program) {
+    ImmutableList<String> localizationStatusNames =
+        localizationUpdate.statuses().stream()
+            .map(LocalizationUpdate.StatusUpdate::configuredStatusText)
+            .collect(ImmutableList.toImmutableList());
+    ImmutableList<String> configuredStatusNames =
+        program.statusDefinitions().getStatuses().stream()
+            .map(StatusDefinitions.Status::statusText)
+            .collect(ImmutableList.toImmutableList());
+    return localizationStatusNames.equals(configuredStatusNames);
+  }
+
   @Override
   public ErrorAnd<ProgramDefinition, CiviFormError> updateLocalization(
-      long programId, Locale locale, String displayName, String displayDescription)
-      throws ProgramNotFoundException {
+      long programId, Locale locale, LocalizationUpdate localizationUpdate)
+      throws ProgramNotFoundException, OutOfDateStatusesException {
     ProgramDefinition programDefinition = getProgramDefinition(programId);
     ImmutableSet.Builder<CiviFormError> errorsBuilder = ImmutableSet.builder();
-    validateProgramText(errorsBuilder, "display name", displayName);
-    validateProgramText(errorsBuilder, "display description", displayDescription);
+    validateProgramText(errorsBuilder, "display name", localizationUpdate.localizedDisplayName());
+    validateProgramText(
+        errorsBuilder, "display description", localizationUpdate.localizedDisplayDescription());
+
+    // It is permissible to set the localized text / email for a status to empty, which falls back
+    // on the English text.
+    if (!localizationStatusesAreInSyncWithConfiguredStatuses(
+        localizationUpdate, programDefinition)) {
+      throw new OutOfDateStatusesException();
+    }
+
+    // We iterate the existing statuses along with the provided statuses since they were verified
+    // to be consistently ordered above.
+    ImmutableList.Builder<StatusDefinitions.Status> toUpdateStatusesBuilder =
+        ImmutableList.builder();
+    for (int statusIdx = 0;
+        statusIdx < programDefinition.statusDefinitions().getStatuses().size();
+        statusIdx++) {
+      LocalizationUpdate.StatusUpdate statusUpdateData =
+          localizationUpdate.statuses().get(statusIdx);
+      StatusDefinitions.Status existingStatus =
+          programDefinition.statusDefinitions().getStatuses().get(statusIdx);
+      StatusDefinitions.Status.Builder updateBuilder =
+          StatusDefinitions.Status.builder()
+              .setStatusText(existingStatus.statusText())
+              .setLocalizedStatusText(
+                  existingStatus
+                      .localizedStatusText()
+                      .updateTranslation(locale, statusUpdateData.localizedStatusText()));
+      if (existingStatus.localizedEmailBodyText().isPresent()) {
```

Comment:
doc this if/else/throw, I think it's

// If the status has email content, update the localization to whatever was provided; otherwise if there's a localization update when there is no email content to localize, that indicates a mismatch between the frontend and the database.

Reply:
Yah, your explanation is correct. Added.